Android/app/src/main/java/app/intra/socks/OverrideSocksHandler.java

```diff
@@ -0,0 +1,366 @@
+
+        listener = splitRetry(upload, download, socketUpload, listener);
+      }
+    } catch (InterruptedException e) {
+      // An interrupt is a normal lifecycle event and indicates that we should perform a clean
+      // shutdown.
+    }
+
+    // Terminate and release both sockets.
+    upload.stop();
+    download.stop();
+    socket.close();
+    session.close();
+
+    if (listener != null) {
+      // We had a functional socket long enough to record statistics.
+      // Report the BYTES event : value = total transfer over the lifetime of a socket
+      // - PORT : TCP port number (i.e. protocol type)
+      // - DURATION: socket lifetime in seconds
+
+      Bundle event = new Bundle();
+      event.putInt(Param.VALUE, listener.uploadBytes + listener.downloadBytes);
+
+      int port = listener.port;
+      if (port >= 0) {
+        if (port != HTTP_PORT && port != HTTPS_PORT && port != 0) {
+          // Group all other ports together into an "other" bucket.
+          port = -1;
+        }
+        event.putInt(Names.PORT.name(), port);
+      }
+
+      int durationMs = listener.getDurationMs();
+      if (durationMs >= 0) {
+        // Socket connection duration is in seconds.
+        event.putInt(Names.DURATION.name(), durationMs / 1000);
+      }
+
+      FirebaseAnalytics.getInstance(context).logEvent(Names.BYTES.name(), event);
+    }
+  }
+
+  /**
+   * This class collects metadata about a proxy socket, including its duration, the amount
+   * transferred, and whether it was terminated normally or by an error.
+   *
+   * It also replicates the wasStopped() and await() functions from Socks5Handler.StopListener.
+   */
+  private static class StatsListener implements PipeListener {
+    // System clock start and stop time for this socket pair.
+    long startTime = -1;
+    long stopTime = -1;
+
+    // If non-null, the socket was closed with an error.
+    Exception error = null;
+
+    // Counters for upload and download total size and number of transfers.
+    int uploadBytes = 0;
+    int uploadCount = 0;
+    int downloadBytes = 0;
+
+    // Server port, for protocol identification.
+    int port;
+
+    boolean simulateReset;
+
+    // The upload buffer is only intended to hold the first flight from the client.
+    private static final int MAX_BUFFER = 1024;
+    private ByteBuffer uploadBuffer = null;
+
+    // Incorporate the functionality of StopListener in Socks5Handler.
+    private final CountDownLatch latch = new CountDownLatch(1);
+    private boolean stopped = false;
+
+    /**
+     * Blocks until onStop is called.
+     * @throws InterruptedException if this thread is interrupted.
+     */
+    void await() throws InterruptedException {
+      latch.await();
+    }
+
+    boolean wasStopped() {
+      return stopped;
+    }
+
+    StatsListener(int port, boolean simulateReset) {
+      this.port = port;
+      this.simulateReset = simulateReset;
+      if (port == HTTPS_PORT) {
```

Comment:
The uploadBuffer is only needed when a SIMULATE_RESET message is going to be sent, correct?  If so, I think the condition required to do ByteBuffer.allocateDirect can be more restrictive than just checking the port.

Reply:
If `SIMULATE_RESET` is false, we still need `uploadBuffer` to be populated in case of an actual TCP RST.